### PR TITLE
Disable qcalls on wasm. Treat them as normal pinvokes.

### PR DESCRIPTION
**Pull request by vargaz**
This is needed because they are stored in a separate table, so they cannot be linked out the same
way as pinvokes/icalls.

### CODE REVIEW

File: src/mono/wasm/wasm.proj [hunk 1]
```diff
@@ -34,6 +34,7 @@
     <ItemGroup>
       <WasmPInvokeModule Include="libSystem.Native" />
       <WasmPInvokeModule Include="libSystem.IO.Compression.Native" />
+      <WasmPInvokeModule Include="QCall" />
```
Comment (marek-safar) [inline src/mono/wasm/wasm.proj:37]: Why is this needed?

File: tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs [hunk 2]
```diff
@@ -154,6 +153,12 @@ private string GenPInvokeDecl(PInvoke pinvoke)
     {
         var sb = new StringBuilder();
         var method = pinvoke.Method;
+        if (method.Name == "EnumCalendarInfo") {
```
Comment (marek-safar) [inline tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs:156]: There is a patch which adds a lot more of these so this won't work

**Comment by marek-safar** [comment]
@lambdageek please review

File: src/mono/wasm/wasm.proj (hunk 1 above)
Comment (vargaz) [reply src/mono/wasm/wasm.proj:37]: Because QCalls are implemented on the managed side as pinvokes with to the module 'QCall'.

File: tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs (hunk 2 above)
Comment (CoffeeFlux) [reply tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs:156]: It's just intended as a temporary hack until that MetadataLoadContext bug is fixed though, so it might be fine for now? Unless you think that patch is landing soon.

File: tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs [hunk 3]
```diff
@@ -154,6 +153,12 @@ private string GenPInvokeDecl(PInvoke pinvoke)
...
+            // FIXME: System.Reflection.MetadataLoadContext can't decode function pointer types
+            // https://github.com/dotnet/runtime/issues/43791
+            sb.Append ($"int {pinvoke.EntryPoint} (int, int, int, int, int);");
```
Comment (CoffeeFlux) [inline tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs:159]:
```suggestion
            sb.Append($"int {pinvoke.EntryPoint} (int, int, int, int, int);");
```

File: tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs [hunk 4]
```diff
@@ -154,6 +153,12 @@ private string GenPInvokeDecl(PInvoke pinvoke)
...
+            return sb.ToString ();
```
Comment (CoffeeFlux) [inline tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs:160]:
```suggestion
            return sb.ToString();
```

**Review by lambdageek — COMMENTED** [review]
I think you also need to exclude `entrypoints.c` from the sources in src/mono/mini/CMakeLists.txt.  Otherwise `gPalGlobalizationNative` will still build the array of all the globalization functions.

**Comment by vargaz** [comment]
> I think you also need to exclude `entrypoints.c` from the sources in src/mono/mini/CMakeLists.txt. Otherwise `gPalGlobalizationNative` will still build the array of all the globalization functions.

It will be built, but nothing will reference it. Will check.

**Comment by vargaz** [comment]
> > I think you also need to exclude `entrypoints.c` from the sources in src/mono/mini/CMakeLists.txt. Otherwise `gPalGlobalizationNative` will still build the array of all the globalization functions.
> 
> It will be built, but nothing will reference it. Will check.

Removing it doesn't seem to matter.

**Review by CoffeeFlux — APPROVED** [review]

**Review by lambdageek — APPROVED** [review]

File: tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs (hunk 2 above)
Comment (marek-safar) [reply tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs:156]: That patch landed yesterday so this does not work anymore

File: src/mono/wasm/wasm.proj (hunk 1 above)
Comment (marek-safar) [reply src/mono/wasm/wasm.proj:37]: Wouldn't be better to handle that inside runtime?

File: tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs (hunk 2 above)
Comment (lambdageek) [reply tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs:156]: catch `NotSupportedException` and map to `int` 😬

File: src/mono/wasm/wasm.proj (hunk 1 above)
Comment (lambdageek) [reply src/mono/wasm/wasm.proj:37]: > Wouldn't be better to handle that inside runtime?

We want to make it possible to remove the native code of pinvokes/qcalls that are not called anywhere in managed.   If the runtime has an array of qcall functions, the native linker won't be able to drop

File: tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs (hunk 2 above)
Comment (lewing) [reply tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs:156]:
```
/Users/lewing/Source/runtime/src/mono/wasm/wasm.proj(48,5): error MSB4018: The "PInvokeTableGenerator" task failed unexpectedly.
/Users/lewing/Source/runtime/src/mono/wasm/wasm.proj(48,5): error MSB4018: System.Exception: The return type 'Interop+BOOL' of pinvoke callback method 'Interop+BOOL EnumCalendarInfoCallback(System.Char*, System.UInt32, System.IntPtr, System.Void*)' needs to be blittable.
/Users/lewing/Source/runtime/src/mono/wasm/wasm.proj(48,5): error MSB4018:    at PInvokeTableGenerator.Error(String msg) in /Users/lewing/Source/runtime/tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs:line 313
/Users/lewing/Source/runtime/src/mono/wasm/wasm.proj(48,5): error MSB4018:    at PInvokeTableGenerator.EmitNativeToInterp(StreamWriter w, List`1 callbacks) in /Users/lewing/Source/runtime/tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs:line 196
/Users/lewing/Source/runtime/src/mono/wasm/wasm.proj(48,5): error MSB4018:    at PInvokeTableGenerator.GenPInvokeTable(String[] pinvokeModules, String[] assemblies) in /Users/lewing/Source/runtime/tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs:line 53
/Users/lewing/Source/runtime/src/mono/wasm/wasm.proj(48,5): error MSB4018:    at PInvokeTableGenerator.Execute() in /Users/lewing/Source/runtime/tools-local/tasks/mobile.tasks/WasmAppBuilder/PInvokeTableGenerator.cs:line 28
/Users/lewing/Source/runtime/src/mono/wasm/wasm.proj(48,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
/Users/lewing/Source/runtime/src/mono/wasm/wasm.proj(48,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask)

```
